ziro-export-risport.ps1

```diff
@@ -0,0 +1,227 @@
+      $devicesToQuery = $DeviceNames[$windowStartIndex..$windowEndIndex]  # upper bound is inclusive
+      $response = $CucmConnector.SelectCmDevice($devicesToQuery)
+      $totalDevicesFound = $response.Envelope.Body.selectCmDeviceResponse.selectCmDeviceReturn.SelectCmDeviceResult.TotalDevicesFound
+
+      if ($totalDevicesFound -gt 0) {
+         $devicesPerNode = $response.Envelope.Body.selectCmDeviceResponse.selectCmDeviceReturn.SelectCmDeviceResult.CmNodes.item.CmDevices 
+         | Where-Object { $_.HasChildNodes }
+
+         foreach ($nodeDevices in $devicesPerNode) {
+            $nodeDevices.item | Select-Object -Property Name, Status, StatusReason, TimeStamp 
+            | ForEach-Object { $registrationStatuses.Add($_.Name, $_) }
+         }
+      }
+
+      $windowStartIndex = $windowStartIndex + $maxReturnedDevices
+      Start-Sleep -Seconds 1  # a pinch of throttling to help with rate limits
+   }
+
+   return $registrationStatuses
+}
+
+
+$serverUrl = Read-Host 'CUCM server URL (https://mycucm.company.com)'
+$credential = Get-Credential -Message 'Enter username and password'
+$cucmConnector = [CucmConnector]::new($serverUrl, $credential) 
+
+try {
+   Write-Host "Fetching devices from CUCM server..."
+   $devices = Get-Devices -CucmConnector $cucmConnector
+   Write-Host "Found [$($devices.Length)] devices"
+   Write-Host "Getting device registration statuses..."
+   $registrationStatuses = Get-DeviceRegistrationStatuses -DeviceNames $devices.name -CucmConnector $cucmConnector
+   $exportResults = New-Object System.Collections.Generic.List[Hashtable]
+
+   foreach ($device in $devices) {
+      if ($registrationStatuses.ContainsKey($device.name)) {
+         $matchingStatus = $registrationStatuses[$device.name]
+         $exportResults.Add(@{
+               Name         = $device.name
+               Status       = $matchingStatus.Status
+               StatusReason = $matchingStatus.StatusReason
+               TimeStamp    = $matchingStatus.TimeStamp
+               Model        = $device.model
+            })
+      }
+      else {
+         $exportResults.Add(@{
+               Name         = $device.name
+               Status       = "Unknown"
+               StatusReason = ""
+               TimeStamp    = ""
+               Model        = $device.model
+            })
+      }
+   }
+
+   # Export to JSON file
+   $serverHost = ([System.Uri]$serverUrl).Host
+   $date = (Get-Date -Format 'yyyy-MM-dd_HH-mm-ss').ToString()
+   $outputFileName = "${date}_${serverHost}.json"
+   $outputFile = $exportResults | ConvertTo-Json -depth 100 -AsArray | New-Item -Path . -Name $outputFileName -ItemType File
+   Write-Host "RISport information export was successful: $outputFile"
+   exit 0
+}
+catch {
+   $responseCode = $_.Exception.Response.StatusCode.value__
+
+   if ($responseCode -eq 401) {
+      Write-Error "Invalid credentials (401 Unauthorized): $_"
```

Comment:
This is what I get when there's a 401:

```
Write-Error: Invalid credentials (401 Unauthorized):                         Cisco System - null                            a {                color: #316984;                text-decoration: none;            }            a:hover {                color: #316984;                text-decoration: underline;            }            a:visted {                color: #316984;                text-decoration: none;            }            body {                background-attachment: fixed;                background-color: #ffffff;                background-repeat: no-repeat;                color: #316984;                font-family: arial,helvetica,sans-serif;            }            #content {                border: 1px solid #d6d7d6;                font-size: 93.5%;                margin: 0px 10% 30px 10%;            }            #content-header {                background-color: #eeeeee;                border-bottom: 1px solid #666666;                color: #666666;                font-size: 124.5%;                padding: 5px 15px 5px 15px;            }            #copyright {                font-size: 75%;                margin: 0px 10% 0px 10%;                padding: 3px 0px 0px 0px;                text-align: right;            }            img {                display: block;                margin: 0px 0px 20px 0px;            }            #logo {                margin: 30px 10% 0px 10%;            }            p {                padding: 5px 15px 5px 15px;            }            pre {                padding: 5px 15px 5px 30px;            }                                                                HTTP Status 401 –                                                             type:                Status Report                                        Message:                                                        Description:                http.401
```

It looks like the API is returning HTML instead of JSON. Technically, the success criterion is fulfilled because the error is appearing on screen, but I'm not sure how useful this is for the user. Maybe it would be best to check with @benoitdesnoyers for this.

Reply:
Talked to Benoit, we decided to not output the error from the API for 401/403s.